django/thunderstore/community/models/community.py

```diff
@@ -206,3 +199,140 @@ def valid_review_statuses(self):
+
+    def as_class(self) -> AggregatedFields:
+        """
+        Return fields in a format supporting attribute access.
+        """
+        return AggregatedFields(self.download_count, self.package_count)
+
+    @classmethod
+    @transaction.atomic
+    def update_for_communities(
+        cls,
+        communities: QuerySet[Community],
+        batch_size: Optional[int] = None,
+    ) -> None:
+        """
+        Updates field values for given Communities.
+
+        Creates the CommunityAggregatedFields object and updates the
+        reference on Community object if none existed previously.
+        """
+        aggregated = cls._aggregate(communities)
+        communities_with_af = [c for c in communities if c.aggregated_fields]
+        communities_without_af = [c for c in communities if not c.aggregated_fields]
+
+        # PrimaryKey and ForeignKey restrictions requires extra steps for
+        # communitites that doesn't have aggregated_fields reference yet.
+        created_afs = cls.objects.bulk_create(
+            [cls.from_dataclass(aggregated[c.id]) for c in communities_without_af],
+            batch_size=batch_size,
+        )
+
+        for community, aggregated_fields in zip(communities_without_af, created_afs):
+            community.aggregated_fields = aggregated_fields
+
+        Community.objects.bulk_update(
+            communities_without_af,
+            ["aggregated_fields"],
+            batch_size,
+        )
+
+        # Update field values for pre-existing CommunityAggregatedFields.
+        updated = [
+            c.aggregated_fields._update_fields(aggregated[c.id])
+            for c in communities_with_af
+        ]
+        fields = ["download_count", "package_count"]
+        cls.objects.bulk_update(updated, fields, batch_size=batch_size)
+
+    @classmethod
+    def _aggregate(
+        cls,
+        communities: QuerySet[Community],
+    ) -> DefaultDict[int, AggregatedFields]:
+        """
+        Precalculates the values for download_count and package_count.
+
+        Helper method for update_for_communities.
+        """
+        communities = communities.prefetch_related(
+            "package_listings__package__versions",
+        )
+        results: DefaultDict[int, AggregatedFields] = defaultdict(
+            lambda: AggregatedFields(),
+        )
+
+        # Manually loop through the results and ignore deactiveted items
+        # etc. to avoid the extra database hits caused by using the
+        # existing convenience methods.
+        for community in communities:
+            if not community.is_listed:
+                # Don't override existing values with the default values
+                # from defaultdict if the community is not processed.
+                if community.aggregated_fields:
+                    results[community.pk] = community.aggregated_fields.as_class()
+
+                continue
+
+            for listing in community.package_listings.all():
```

Comment:
This is probably not a good idea in terms of memory usage. I see two realistic alternatives:
1. We stop caring about metrics being in line with the more complex visibility rules and just filter for `is_active` in the DB query, then aggregate
2. We use https://docs.djangoproject.com/en/4.2/ref/models/querysets/#iterator. Note that this breaks `prefetch_related`

I'd vote for option 1

Reply:
Not sure if I understood the comment correctly, but I ended up doing nothing for this, since I'm not sure if this comment is still valid now that each community is handled separately.